File: libs/common/estategap_common/models/listing.py
"""Listing and price-history Pydantic models."""

from __future__ import annotations

from datetime import date
from decimal import Decimal
from enum import Enum
from typing import Any, Literal
from uuid import UUID

from pydantic import Field, field_validator

from ._base import AwareDatetime, EstateGapModel, validate_country_code, validate_currency_code
from .scoring import DealTier, ShapValue

EnrichmentState = Literal["pending", "completed", "partial", "no_match", "failed"]


class PropertyCategory(str, Enum):
    """Supported listing category values shared across services."""

    RESIDENTIAL = "residential"
    COMMERCIAL = "commercial"
    INDUSTRIAL = "industrial"
    LAND = "land"


class ListingStatus(str, Enum):
    """Persisted listing lifecycle states."""

    ACTIVE = "active"
    DELISTED = "delisted"
    SOLD = "sold"


class RawListing(EstateGapModel):
    """Raw listing payload emitted by a scraper before normalization."""

    external_id: str
    portal: str
    country_code: str
    raw_json: dict[str, Any]
    scraped_at: AwareDatetime

    @field_validator("country_code")
    @classmethod
    def _validate_country_code(cls, value: str) -> str:
        return validate_country_code(value)


class NormalizedListing(EstateGapModel):
    """Validated listing data ready for database insertion."""

    id: UUID
    canonical_id: UUID | None = None
    country: str
    source: str
    source_id: str
    source_url: str
    portal_id: UUID | None = None
    address: str | None = None
    city: str | None = None
    region: str | None = None
    postal_code: str | None = None
    location_wkt: str | None = None
    zone_id: UUID | None = None
    asking_price: Decimal
    currency: str
    asking_price_eur: Decimal
    price_per_m2_eur: Decimal | None = None
    property_category: PropertyCategory | None = None
    property_type: str | None = None
    built_area_sqft: Decimal | None = None
    built_area_m2: Decimal
    usable_area_m2: Decimal | None = None
    plot_area_m2: Decimal | None = None
    lot_size_sqft: Decimal | None = None
    lot_size_m2: Decimal | None = None
    bedrooms: int | None = None
    bathrooms: int | None = None
    floor_number: int | None = None
    total_floors: int | None = None
    parking_spaces: int | None = None
    has_lift: bool | None = None
    has_pool: bool | None = None
    year_built: int | None = None
    council_tax_band: str | None = None
    epc_rating: str | None = None
    tenure: str | None = None
    leasehold_years_remaining: int | None = None
    seller_type: str | None = None
    omi_zone_code: str | None = None
    omi_price_min_eur_m2: Decimal | None = None
    omi_price_max_eur_m2: Decimal | None = None
    omi_period: str | None = None
    price_vs_omi: Decimal | None = None
    bag_id: str | None = None
    official_area_m2: Decimal | None = None
    dvf_nearby_count: int | None = None
    dvf_median_price_m2: Decimal | None = None
    uk_lr_match_count: int | None = None
    uk_lr_last_price_gbp: int | None = None
    uk_lr_last_date: date | None = None
    cadastral_ref: str | None = None
    official_built_area_m2: Decimal | None = None
    area_discrepancy_flag: bool | None = None
    building_geometry_wkt: str | None = None
    enrichment_status: EnrichmentState | None = None
    enrichment_attempted_at: AwareDatetime | None = None
    dist_metro_m: int | None = None
    dist_train_m: int | None = None
    dist_airport_m: int | None = None
    dist_park_m: int | None = None
    dist_beach_m: int | None = None
    hoa_fees_monthly_usd: int | None = None
    tax_assessed_value_usd: int | None = None
    school_rating: Decimal | None = None
    zestimate_reference_usd: int | None = None
    compete_score: int | None = None
    mls_id: str | None = None
    condition: str | None = None
    energy_rating: str | None = None
    status: ListingStatus = ListingStatus.ACTIVE
    description_orig: str | None = None
    images_count: int = 0
    data_completeness: float | None = None
    first_seen_at: AwareDatetime
    last_seen_at: AwareDatetime
    published_at: AwareDatetime | None = None
    raw_hash: str | None = None

    @field_validator("country")
    @classmethod
    def _validate_country(cls, value: str) -> str:
        return validate_country_code(value)

    @field_validator("currency")
    @classmethod
    def _validate_currency(cls, value: str) -> str:
        return validate_currency_code(value)

    @field_validator("asking_price", "built_area_m2")
    @classmethod
    def _validate_positive_decimals(cls, value: Decimal, info: Any) -> Decimal:
        if value <= 0:
            raise ValueError(f"{info.field_name} must be greater than zero")
        return value


class Listing(NormalizedListing):
    """Full persisted listing record shared by the API, pipeline, and ML layers."""

    neighborhood: str | None = None
    district: str | None = None
    toilets: int | None = None
    has_garden: bool | None = None
    terrace_area_m2: Decimal | None = None
    garage_area_m2: Decimal | None = None
    last_renovated: int | None = None
    energy_rating_kwh: Decimal | None = None
    co2_rating: str | None = None
    co2_kg_m2: Decimal | None = None
    frontage_m: Decimal | None = None
    ceiling_height_m: Decimal | None = None
    loading_docks: int | None = None
    power_kw: Decimal | None = None
    office_area_m2: Decimal | None = None
    warehouse_area_m2: Decimal | None = None
    buildability_index: Decimal | None = None
    urban_classification: str | None = None
    land_use: str | None = None
    estimated_price: Decimal | None = None
    deal_score: Decimal | None = None
    deal_tier: DealTier | None = None
    confidence_low: Decimal | None = None
    confidence_high: Decimal | None = None
    shap_features: list[ShapValue] | None = None
    comparable_ids: list[UUID] | None = None
    model_version: str | None = None
    scored_at: AwareDatetime | None = None
    days_on_market: int | None = None
    description_lang: str | None = None
    delisted_at: AwareDatetime | None = None
    created_at: AwareDatetime
    updated_at: AwareDatetime


class PriceHistory(EstateGapModel):
    """A single listing price-change event."""

    id: int
    listing_id: UUID
    country: str
    old_price: Decimal | None = None
    new_price: Decimal
    currency: str
    old_price_eur: Decimal | None = None
    new_price_eur: Decimal | None = None
    change_type: str = "price_change"
    old_status: ListingStatus | None = None
    new_status: ListingStatus | None = None
    recorded_at: AwareDatetime
    source: str | None = None

    @field_validator("country")
    @classmethod
    def _validate_country(cls, value: str) -> str:
        return validate_country_code(value)

    @field_validator("currency")
    @classmethod
    def _validate_currency(cls, value: str) -> str:
        return validate_currency_code(value)


PriceChange = PriceHistory


class PriceChangeEvent(EstateGapModel):
    """Price-drop event emitted for downstream consumers."""

    listing_id: UUID
    country: str
    portal: str
    old_price: Decimal
    new_price: Decimal
    currency: str
    old_price_eur: Decimal | None = None
    new_price_eur: Decimal | None = None
    drop_percentage: Decimal
    recorded_at: AwareDatetime

    @field_validator("country")
    @classmethod
    def _validate_country(cls, value: str) -> str:
        return validate_country_code(value)

    @field_validator("currency")
    @classmethod
    def _validate_currency(cls, value: str) -> str:
        return validate_currency_code(value)


class ScrapeCycleEvent(EstateGapModel):
    """Cycle-completion event emitted by the scrape orchestrator."""

    cycle_id: str
    portal: str
    country: str
    completed_at: AwareDatetime
    listing_ids: list[str] = Field(default_factory=list)

    @field_validator("country")
    @classmethod
    def _validate_country(cls, value: str) -> str:
        return validate_country_code(value)


__all__ = [
    "EnrichmentState",
    "Listing",
    "ListingStatus",
    "NormalizedListing",
    "PriceChange",
    "PriceChangeEvent",
    "PriceHistory",
    "PropertyCategory",
    "RawListing",
    "ScrapeCycleEvent",
]

File: libs/pkg/models/listing.go
package models

import (
	"encoding/json"

	"github.com/jackc/pgx/v5/pgtype"
	"github.com/shopspring/decimal"
)

type Listing struct {
	ID                  pgtype.UUID         `json:"id" db:"id"`
	CanonicalID         *pgtype.UUID        `json:"canonical_id" db:"canonical_id"`
	Country             string              `json:"country" db:"country"`
	Source              string              `json:"source" db:"source"`
	SourceID            string              `json:"source_id" db:"source_id"`
	SourceURL           string              `json:"source_url" db:"source_url"`
	PortalID            *pgtype.UUID        `json:"portal_id" db:"portal_id"`
	Address             *string             `json:"address" db:"address"`
	Neighborhood        *string             `json:"neighborhood" db:"neighborhood"`
	District            *string             `json:"district" db:"district"`
	City                *string             `json:"city" db:"city"`
	Region              *string             `json:"region" db:"region"`
	PostalCode          *string             `json:"postal_code" db:"postal_code"`
	LocationWKT         *string             `json:"location_wkt,omitempty" db:"location_wkt"`
	ZoneID              *pgtype.UUID        `json:"zone_id" db:"zone_id"`
	AskingPrice         *decimal.Decimal    `json:"asking_price" db:"asking_price"`
	Currency            string              `json:"currency" db:"currency"`
	AskingPriceEUR      *decimal.Decimal    `json:"asking_price_eur" db:"asking_price_eur"`
	PriceConverted      *decimal.Decimal    `json:"price_converted,omitempty" db:"price_converted"`
	PricePerM2EUR       *decimal.Decimal    `json:"price_per_m2_eur" db:"price_per_m2_eur"`
	PropertyCategory    *PropertyCategory   `json:"property_category" db:"property_category"`
	PropertyType        *string             `json:"property_type" db:"property_type"`
	BuiltAreaM2         *decimal.Decimal    `json:"built_area_m2" db:"built_area_m2"`
	UsableAreaM2        *decimal.Decimal    `json:"usable_area_m2" db:"usable_area_m2"`
	PlotAreaM2          *decimal.Decimal    `json:"plot_area_m2" db:"plot_area_m2"`
	Bedrooms            *int16              `json:"bedrooms" db:"bedrooms"`
	Bathrooms           *int16              `json:"bathrooms" db:"bathrooms"`
	Toilets             *int16              `json:"toilets,omitempty" db:"toilets"`
	FloorNumber         *int16              `json:"floor_number,omitempty" db:"floor_number"`
	TotalFloors         *int16              `json:"total_floors,omitempty" db:"total_floors"`
	ParkingSpaces       *int16              `json:"parking_spaces,omitempty" db:"parking_spaces"`
	HasLift             *bool               `json:"has_lift,omitempty" db:"has_lift"`
	HasPool             *bool               `json:"has_pool,omitempty" db:"has_pool"`
	HasGarden           *bool               `json:"has_garden,omitempty" db:"has_garden"`
	TerraceAreaM2       *decimal.Decimal    `json:"terrace_area_m2,omitempty" db:"terrace_area_m2"`
	GarageAreaM2        *decimal.Decimal    `json:"garage_area_m2,omitempty" db:"garage_area_m2"`
	YearBuilt           *int16              `json:"year_built,omitempty" db:"year_built"`
	LastRenovated       *int16              `json:"last_renovated,omitempty" db:"last_renovated"`
	Condition           *string             `json:"condition,omitempty" db:"condition"`
	EnergyRating        *string             `json:"energy_rating,omitempty" db:"energy_rating"`
	EnergyRatingKWH     *decimal.Decimal    `json:"energy_rating_kwh,omitempty" db:"energy_rating_kwh"`
	CO2Rating           *string             `json:"co2_rating,omitempty" db:"co2_rating"`
	CO2KgM2             *decimal.Decimal    `json:"co2_kg_m2,omitempty" db:"co2_kg_m2"`
	FrontageM           *decimal.Decimal    `json:"frontage_m,omitempty" db:"frontage_m"`
	CeilingHeightM      *decimal.Decimal    `json:"ceiling_height_m,omitempty" db:"ceiling_height_m"`
	LoadingDocks        *int16              `json:"loading_docks,omitempty" db:"loading_docks"`
	PowerKW             *decimal.Decimal    `json:"power_kw,omitempty" db:"power_kw"`
	OfficeAreaM2        *decimal.Decimal    `json:"office_area_m2,omitempty" db:"office_area_m2"`
	WarehouseAreaM2     *decimal.Decimal    `json:"warehouse_area_m2,omitempty" db:"warehouse_area_m2"`
	BuildabilityIndex   *decimal.Decimal    `json:"buildability_index,omitempty" db:"buildability_index"`
	UrbanClassification *string             `json:"urban_classification,omitempty" db:"urban_classification"`
	LandUse             *string             `json:"land_use,omitempty" db:"land_use"`
	EstimatedPrice      *decimal.Decimal    `json:"estimated_price" db:"estimated_price_eur"`
	DealScore           *decimal.Decimal    `json:"deal_score" db:"deal_score"`
	DealTier            *DealTier           `json:"deal_tier" db:"deal_tier"`
	ConfidenceLow       *decimal.Decimal    `json:"confidence_low" db:"confidence_low_eur"`
	ConfidenceHigh      *decimal.Decimal    `json:"confidence_high" db:"confidence_high_eur"`
	ShapFeatures        json.RawMessage     `json:"shap_features" db:"shap_features"`
	ComparableIDs       []pgtype.UUID       `json:"comparable_ids,omitempty" db:"comparable_ids"`
	ModelVersion        *string             `json:"model_version" db:"model_version"`
	ScoredAt            *pgtype.Timestamptz `json:"scored_at" db:"scored_at"`
	DaysOnMarket        *int32              `json:"days_on_market" db:"days_on_market"`
	Status              ListingStatus       `json:"status" db:"status"`
	DescriptionOrig     *string             `json:"description_orig,omitempty" db:"description_orig"`
	DescriptionLang     *string             `json:"description_lang,omitempty" db:"description_lang"`
	ImagesCount         int16               `json:"images_count" db:"images_count"`
	FirstSeenAt         pgtype.Timestamptz  `json:"first_seen_at" db:"first_seen_at"`
	ExchangeRateDate    pgtype.Date         `json:"-" db:"exchange_rate_date"`
	LastSeenAt          pgtype.Timestamptz  `json:"last_seen_at" db:"last_seen_at"`
	PublishedAt         *pgtype.Timestamptz `json:"published_at" db:"published_at"`
	DelistedAt          *pgtype.Timestamptz `json:"delisted_at" db:"delisted_at"`
	RawHash             *string             `json:"raw_hash,omitempty" db:"raw_hash"`
	CreatedAt           pgtype.Timestamptz  `json:"created_at" db:"created_at"`
	UpdatedAt           pgtype.Timestamptz  `json:"updated_at" db:"updated_at"`
}

type PriceHistory struct {
	ID          int64              `json:"id" db:"id"`
	ListingID   pgtype.UUID        `json:"listing_id" db:"listing_id"`
	Country     string             `json:"country" db:"country"`
	OldPrice    *decimal.Decimal   `json:"old_price" db:"old_price"`
	NewPrice    decimal.Decimal    `json:"new_price" db:"new_price"`
	Currency    string             `json:"currency" db:"currency"`
	OldPriceEUR *decimal.Decimal   `json:"old_price_eur" db:"old_price_eur"`
	NewPriceEUR *decimal.Decimal   `json:"new_price_eur" db:"new_price_eur"`
	ChangeType  string             `json:"change_type" db:"change_type"`
	OldStatus   *ListingStatus     `json:"old_status" db:"old_status"`
	NewStatus   *ListingStatus     `json:"new_status" db:"new_status"`
	RecordedAt  pgtype.Timestamptz `json:"recorded_at" db:"recorded_at"`
	Source      *string            `json:"source" db:"source"`
}
